src/FrameDump.cpp

```diff
@@ -0,0 +1,209 @@
+    Platform::FileWrite(GPU.GPU3D.RenderToonTable, 1, sizeof(GPU.GPU3D.RenderToonTable), file);
+
+    // save initial state vars to file.
+    Platform::FileWrite(&ZDotDisp_Track, 1, 1, file);
+    Platform::FileWrite(&ZDotDisp, 1, sizeof(ZDotDisp), file);
+    Platform::FileWrite(&PolyAttr, 1, sizeof(PolyAttr), file);
+    Platform::FileWrite(&PolyAttrUnset, 1, sizeof(PolyAttrUnset), file);
+    Platform::FileWrite(&VtxColor, 1, sizeof(VtxColor), file);
+    Platform::FileWrite(&Viewport, 1, sizeof(Viewport), file);
+    Platform::FileWrite(ProjStack, 1, sizeof(ProjStack), file);
+    Platform::FileWrite(PosStack, 1, sizeof(PosStack), file);
+    Platform::FileWrite(VecStack, 1, sizeof(VecStack), file);
+    Platform::FileWrite(TexStack, 1, sizeof(TexStack), file);
+    Platform::FileWrite(ProjMtx, 1, sizeof(ProjMtx), file);
+    Platform::FileWrite(PosMtx, 1, sizeof(PosMtx), file);
+    Platform::FileWrite(VecMtx, 1, sizeof(VecMtx), file);
+    Platform::FileWrite(TexMtx, 1, sizeof(TexMtx), file);
+    Platform::FileWrite(&MatrixMode, 1, sizeof(MatrixMode), file);
+    Platform::FileWrite(&Polygon, 1, sizeof(Polygon), file);
+    Platform::FileWrite(&VtxXY, 1, sizeof(VtxXY), file);
+    Platform::FileWrite(&VtxZ, 1, sizeof(VtxZ), file);
+    Platform::FileWrite(&TexCoord, 1, sizeof(TexCoord), file);
+    Platform::FileWrite(&TexParam, 1, sizeof(TexParam), file);
+    Platform::FileWrite(&TexPalette, 1, sizeof(TexPalette), file);
+    Platform::FileWrite(&DiffAmbi, 1, sizeof(DiffAmbi), file);
+    Platform::FileWrite(&SpecEmis, 1, sizeof(SpecEmis), file);
+    Platform::FileWrite(Shininess, 1, sizeof(Shininess), file);
+    Platform::FileWrite(LightVec, 1, sizeof(LightVec), file);
+    Platform::FileWrite(LightColor, 1, sizeof(LightColor), file);
+    Platform::FileWrite(&SwapBuffer, 1, sizeof(SwapBuffer), file);
+    
+    // skip banks H and I, we only care about dumping 3d engine state currently, and banks H and I cannot be used by the 3d engine
+    // save vram control regs to file
+    Platform::FileWrite(GPU.VRAMCNT, 1, sizeof(GPU.VRAMCNT[0])*7, file);
+
+    // only save vram if the bank is enabled and allocated to be used as texture image/texture palette.
+    if ((GPU.VRAMCNT[0] & 0x83) == 0x83) Platform::FileWrite(GPU.VRAM_A, 1, sizeof(GPU.VRAM_A), file);
+    if ((GPU.VRAMCNT[1] & 0x83) == 0x83) Platform::FileWrite(GPU.VRAM_B, 1, sizeof(GPU.VRAM_B), file);
+    if ((GPU.VRAMCNT[2] & 0x87) == 0x83) Platform::FileWrite(GPU.VRAM_C, 1, sizeof(GPU.VRAM_C), file);
+    if ((GPU.VRAMCNT[3] & 0x87) == 0x83) Platform::FileWrite(GPU.VRAM_D, 1, sizeof(GPU.VRAM_D), file);
+    if ((GPU.VRAMCNT[4] & 0x87) == 0x83) Platform::FileWrite(GPU.VRAM_E, 1, sizeof(GPU.VRAM_E), file);
+    if ((GPU.VRAMCNT[5] & 0x87) == 0x83) Platform::FileWrite(GPU.VRAM_F, 1, sizeof(GPU.VRAM_F), file);
+    if ((GPU.VRAMCNT[6] & 0x87) == 0x83) Platform::FileWrite(GPU.VRAM_G, 1, sizeof(GPU.VRAM_G), file);
+    
+    u8* tempcmds = (u8*)malloc(Cmds.size() * sizeof(u8));
```

Comment:
Personally, I'd use a `std::vector` here instead of `malloc`, possibly initialized to a known size.

Reply:
i figured malloc would work just fine for this situation, since i don't need to do any complex management of the memory, so the potential for a memory leak is low, and the upper bounds of how much memory i'd need in this situation are known, so the ability to dynamically resize seemed unneeded.
Though a dynamically sized array would be slightly more efficient for memory usage since im not guaranteed to be using the full size of the allocated memory, but i figured that hyper optimizing for that didn't really matter.
I can change it if you feel it'd be better though.